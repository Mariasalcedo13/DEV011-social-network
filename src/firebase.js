import {
  deleteDoc,
  query,
  where,
  getFirestore,
  collection,
  getDocs,
  addDoc,
  doc,
  getDoc,
  updateDoc,
  onSnapshot,
} from 'firebase/firestore';
import {
  getAuth,
  GoogleAuthProvider,
  signInWithPopup,
  signInWithEmailAndPassword,
  createUserWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
} from 'firebase/auth';
import {
  ref,
  uploadBytes,
  getDownloadURL,
  getStorage,
} from 'firebase/storage';
import { initializeApp } from 'firebase/app';

// import renderCreateAccount from './register.js';
// import { posts } from './post.js';
// import navigateTo from './main';

const firebaseConfig = {
  apiKey: 'AIzaSyA_nNPVRwXqmgLlxdYL4NmJwiItX9t2D5E',
  authDomain: 'social-network-c61c9.firebaseapp.com',
  projectId: 'social-network-c61c9',
  storageBucket: 'social-network-c61c9.appspot.com',
  messagingSenderId: '496904934051',
  appId: '1:496904934051:web:349b4f181faf09491c2516',
  measurementId: 'G-9MSK8FV9VP',
};
// Initialize Firebase
const app = initializeApp(firebaseConfig);
const auth = getAuth(app);
const firestore = getFirestore(app);
const googleProvider = new GoogleAuthProvider();
googleProvider.setCustomParameters({
  prompt: 'select_account',
});
const storage = getStorage(app);
export {
  app, firestore, googleProvider, auth,
};

// Para crear o registrar usuarios
export function createUser(email, password) {
  return new Promise((resolve, reject) => {
    createUserWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        console.log('Registro exitoso', userCredential);
        const user = userCredential.user;
        resolve({ message: 'success', email: user.email });
      })
      .catch((error) => {
        console.error(
          'Error al registrarse:',
          error.code,
          error.message,
          error.serverResponse,
        );
        error.email = email;
        reject(error);
      });
  });
}

// Para iniciar sesion o ingresar
export function login(email, password) {
  return new Promise((resolve, reject) => {
    signInWithEmailAndPassword(auth, email, password)
      .then((userCredential) => {
        console.log('Inicio sesion exitoso', userCredential);
        const user = userCredential.user;
        resolve({ message: 'success', email: user.email });
      })
      .catch((error) => {
        console.error(
          'Error al iniciar sesion:',
          error.code,
          error.message,
          error.serverResponse,
        );
        error.email = email;
        reject(error);
      });
  });
}

// function save
// export function saveTask(title, description, imageInput) {

//   const imageName = `${Date.now()}_${imageInput.name}`;
//   const storageRef = ref(storage, `/img/${imageName}`);
//   uploadBytes(storageRef, imageInput).then((snapshot) => {
//     console.log('Uploaded a blob or file!');
//   });
//   console.log(storageRef);
//   let image = {};
//   const postCollection = collection(firestore, 'post');
//   addDoc(postCollection, {
//     title,
//     description,
//     image,
//     likes: 0,
//   })
//     .then((docRef) => {
//       console.log('Documento guardado con ID:', docRef.id);
//     })
//     .catch((error) => {
//       console.error('Error al guardar el documento:', error);
//     });
// }
// No necesitas el async/await aquí
export async function saveTask(title, description, imageFile) {
  console.log(imageFile);
  if (!imageFile) {
    const postCollection = collection(firestore, 'post');
    // Guarda la información del post en la base de datos
    try {
      await addDoc(postCollection, {
        title,
        description,
        likes: 0,
      });
    } catch (e) {
      console.error(e);
    }
  } else {
    try {
      const imageName = `${Date.now()}_${imageFile.name}`;
      const storageRef = ref(storage, `/img/${imageName}`);
      // Sube la imagen al almacenamiento y obtiene la URL de descarga
      const imageUrl = await uploadBytes(storageRef, imageFile)
        .then(() => getDownloadURL(storageRef));
      console.log('URL de la imagen:', imageUrl);
      const postCollection = collection(firestore, 'post');
      // Guarda la información del post en la base de datos
      const docRef = await addDoc(postCollection, {
        title,
        description,
        imageUrl,
        likes: 0,
      });
      console.log('Documento guardado con ID:', docRef.id);
    } catch (error) {
      console.error('Error al guardar el documento:', error);
      throw error; // Puedes manejar el error según tus necesidades
    }
  }
}

// funcion para registro con google
export function GoogleRegister(navigateTo) {
  signInWithPopup(auth, googleProvider)
    .then((result) => {
      const user = result.user;
      console.log('Usuario autenticado con Google:', user);
      navigateTo('/posts');
    })
    .catch((error) => {
      console.error('Error al autenticar con Google:', error.message);
    });
}

// funcion para likes
export function handleLike(postId, userId, callback) {
  const likesCollection = collection(firestore, 'likes');
  const likeButton = document.querySelector(
    `.likeButton[data-post-id="${postId}"]`,
  );
  const likeQuery = query(
    likesCollection,
    where('postId', '==', postId),
    where('userId', '==', userId),
  );

  // Ejecuta la consulta para verificar si el usuario ya ha dado like
  getDocs(likeQuery)
    .then((querySnapshot) => {
      // Verifica si no hay likes existentes del usuario para esta publicación
      if (querySnapshot.empty) {
        const postRef = doc(firestore, 'post', postId);

        // Obtiene la información actual de la publicación
        getDoc(postRef)
          .then((docSnapshot) => {
            if (docSnapshot.exists()) {
              const currentLikes = docSnapshot.data().likes;
              const updatedLikes = currentLikes + 1;
              likeButton.style.background = 'green';
              // Actualiza la cantidad de likes en la publicación
              updateDoc(postRef, { likes: updatedLikes })
                .then(() => {
                  console.log('Like registrado con éxito.');

                  // Agrega un nuevo documento de like
                  addDoc(likesCollection, { postId, userId })
                    .then(() => {
                      console.log('Nuevo like registrado con éxito.');
                      callback(); // Llama al callback después de manejar el like
                    })
                    .catch((error) => {
                      console.error('Error al agregar nuevo like:', error);
                    });
                })
                .catch((error) => {
                  console.error('Error al actualizar likes:', error);
                });
            }
          })
          .catch((error) => {
            console.error('Error al obtener documento de publicación:', error);
          });
      } else {
        // El usuario ya ha dado like antes, manejar según sea necesario
        const likeDoc = querySnapshot.docs[0]; // Obtén el documento de like existente
        const likeId = likeDoc.id;

        // Elimina el like existente
        deleteDoc(doc(likesCollection, likeId))
          .then(() => {
            console.log('Like eliminado con éxito.');

            // Obtiene la información actual de la publicación
            const postRef = doc(firestore, 'post', postId);
            getDoc(postRef)
              .then((docSnapshot) => {
                if (docSnapshot.exists()) {
                  const currentLikes = docSnapshot.data().likes;
                  const updatedLikes = currentLikes - 1;

                  // Actualiza la cantidad de likes en la publicación
                  updateDoc(postRef, { likes: updatedLikes })
                    .then(() => {
                      console.log('Like actualizado con éxito.');
                      callback(); // Llama al callback después de manejar el unlike
                    })
                    .catch((error) => {
                      console.error('Error al actualizar likes:', error);
                    });
                }
              })
              .catch((error) => {
                console.error(
                  'Error al obtener documento de publicación:',
                  error,
                );
              });
          })
          .catch((error) => {
            console.error('Error al eliminar like:', error);
          });
      }
    })
    .catch((error) => {
      console.error('Error al realizar la consulta de likes:', error);
    });
}

// para eliminar post
export function deletePost(postId) {
  const postCollection = collection(firestore, 'post');
  const postDocRef = doc(postCollection, postId);
  deleteDoc(postDocRef)
    .then(() => {
      console.log('Post eliminado con éxito.');
    })
    .catch((error) => {
      console.error('Error al borrar post:', error);
    });
}

// para editar los post
export function editPost(postId, updatedTitle, updatedDescription) {
  const postRef = doc(firestore, 'post', postId);

  return updateDoc(postRef, {
    title: updatedTitle,
    description: updatedDescription,
  });
}

// funcion cerrar sesion
export function logOut(navigateTo) {
  signOut(auth)
    .then(() => {
      // Sign-out successful.
      console.error('sesion cerrada');
      navigateTo('/login');
    })
    .catch((error) => {
      console.error('Error al cerrar sesion:', error.code, error.message);
      // An error happened.
    });
}

export function initializeAuth(setupPost, navigateTo) {
  onAuthStateChanged(auth, (user) => {
    if (user) {
      // Usuario autenticado, puedes acceder a la colección de 'post'
      console.log(
        'User authenticated:',
        auth.currentUser.uid,
        'email:',
        user.email,
      );
      const userPostsCollection = collection(firestore, 'post');
      onSnapshot(userPostsCollection, (snapshot) => {
        const postSnap = [];
        snapshot.forEach((docu) => {
          postSnap.push(docu);
        });
        setupPost(postSnap);
      });
    } else {
      console.log('Usuario no autenticado');
      // logOut();
      alert('Usuario no autenticado');
      navigateTo('/login');
      // navigateTo('/login');
      // backgroundLayer.style.opacity = 0.0;
      // Resto del código de la lógica que quieras tras un usuario no autenticado
    }
  });
}
