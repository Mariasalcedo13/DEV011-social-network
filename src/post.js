import { collection, getDocs } from 'firebase/firestore';
import {
  auth,
  firestore,
  saveTask,
  handleLike,
  deletePost,
  editPost,
  logOut,
  initializeAuth,
} from './firebase.js';

export function posts(navigateTo) {
  // const body1 = document.querySelector('body');
  const backgroundLayer = document.createElement('div');
  backgroundLayer.classList.add('background-layer');
  // homepage.style.boxShadow = '0px 0px 0px transparent';
  // homepage.style.height = '100%';
  // homepage.style.width = '100%';
  // homepage.style.paddingTop = '0em';
  // backgroundLayer.style.background = "url('/img/patron2.avif')";
  // backgroundLayer.style.opacity = 0.1;
  // backgroundLayer.style.zIndex = '-1';
  // backgroundLayer.style.top = '0';
  // backgroundLayer.style.left = '0';
  // backgroundLayer.style.width = '100%';
  // backgroundLayer.style.height = '100%';

  const mainPage = document.createElement('div');
  mainPage.setAttribute('class', 'homepage');
  mainPage.setAttribute('id', 'homepagePostes');
  const headerPost = document.createElement('div');
  headerPost.textContent = 'Mi Plantapp';
  headerPost.setAttribute('class', 'headerPost');
  const logoImage = document.createElement('img');
  logoImage.setAttribute('src', '/img/planta-arana.png');
  logoImage.setAttribute('class', 'logoImage');
  // Boton cerrar sesion
  const logOutIcon = document.createElement('button');
  logOutIcon.setAttribute('class', 'logOutButton');
  // icono cerrar sesion
  const iconLogOut = document.createElement('img');
  iconLogOut.setAttribute('src', '/img/salir.png');
  // Contenedor de Creacion de post
  const containerPubication = document.createElement('div');
  containerPubication.setAttribute('class', 'containerPubication');
  // Imagen Post
  const imagePublication = document.createElement('img');
  imagePublication.setAttribute('src', '/img/mujer.png');
  imagePublication.setAttribute('class', 'imagePublication');
  // Formulario para la creacion de post
  const containerPost = document.createElement('form');
  containerPost.setAttribute('id', 'task-form');
  // Input titulo
  const postTitle = document.createElement('input');
  postTitle.setAttribute('type', 'text');
  postTitle.setAttribute('class', 'postTitle');
  postTitle.setAttribute('placeholder', 'Título de la publicación.');
  // Input descripcion
  const post = document.createElement('textarea');
  post.setAttribute('placeholder', 'Ingresa el contenido de la publicación.');
  post.setAttribute('id', 'postText');
  // Nuevo campo para cargar archivos (fotos)
  const imageInput = document.createElement('input');
  imageInput.setAttribute('type', 'file');
  imageInput.setAttribute('id', 'post-image');
  imageInput.setAttribute('accept', 'image/*');
  // Boton publicar
  const buttonSave = document.createElement('button');
  buttonSave.setAttribute('class', 'buttonSave');
  buttonSave.textContent = 'Publicar';
  // contenedor post
  const viewPost = document.createElement('div');
  viewPost.setAttribute('class', 'postView');
  buttonSave.addEventListener('click', (e) => {
    e.preventDefault();
    const title = postTitle.value;
    const description = post.value;
    const imageFile = imageInput.files[0];
    // console.log(title, description);
    if (title === '' || description === '') {
      alert('Campos vacios');
    } else {
      saveTask(title, description, imageFile);
      containerPost.reset();
    }
  });

  function setupPost(data) {
    // console.log('Data inside setupPost:', data);
    if (data) {
      let html = '';
      data.forEach((doc) => {
        const postdata = doc.data();
        if (!postdata.imageUrl) {
          html += `
          <li class="ListGroupItem">
          <div class='buttonOptions'>
          <button class='deleteButton' data-post-id="${doc.id}"> Delete </button>
          <button class='editButton' data-post-id="${doc.id}"> Editar </button>
          </div>
          <h5>${postdata.title}</h5>
          <p>${postdata.description}</p>
          <div class="containerLikes" data-post-id="${doc.id}">
          <button class="likeButton" data-post-id="${doc.id}">
          <img src="img/like.png" class='imgLike'>
          </button>
          <span>${postdata.likes} Likes</span>
          </div>
          <h4 class='editPublic' data-post-id="${doc.id}" style="display: none;"> Editar publicación: </h4>
          <textarea class="editTextarea" data-post-id="${doc.id}" style="display: none;">${postdata.title}</textarea>
          <textarea class="editContentTextarea" data-post-id="${doc.id}" style="display: none;">${postdata.description}</textarea>
          <button class="saveEditButton" data-post-id="${doc.id}" style="display: none;">Guardar</button>
          </li>
          `;
        } else {
          html += `
          <li class="ListGroupItem">
          <div class='buttonOptions'>
          <button class='deleteButton' data-post-id="${doc.id}"> Delete </button>
          <button class='editButton' data-post-id="${doc.id}"> Editar </button>
          </div>
          <h5>${postdata.title}</h5>
          <p>${postdata.description}</p>
          <img src="${postdata.imageUrl}" alt="Imagen del post" data-post-id="${doc.id}" class="imgPostPublication">
          <div class="containerLikes" data-post-id="${doc.id}">
          <button class="likeButton" data-post-id="${doc.id}">
          <img src="img/like.png" class='imgLike'>
          </button>
          <span>${postdata.likes} Likes</span>
          </div>
          <h4 class='editPublic' data-post-id="${doc.id}" style="display: none;"> Editar publicación: </h4>
          <textarea class="editTextarea" data-post-id="${doc.id}" style="display: none;">${postdata.title}</textarea>
          <textarea class="editContentTextarea" data-post-id="${doc.id}" style="display: none;">${postdata.description}</textarea>
          <button class="saveEditButton" data-post-id="${doc.id}" style="display: none;">Guardar</button>
          </li>
          `;
        }
      });
      viewPost.innerHTML = html;
      // Evento Like
      const likeButtons = document.querySelectorAll('.likeButton');
      likeButtons.forEach((button) => {
        button.addEventListener('click', (e) => {
          const postId = e.currentTarget.getAttribute('data-post-id');
          const userId = auth.currentUser.uid; // Obtiene el id del usuario actual
          const isLiked = e.currentTarget.classList.contains('liked'); // Verifica si el botón ya ha sido "liked"
          // Cambiar el estado del botón "like" (colorear o quitar el color rojo)
          if (isLiked) {
            e.currentTarget.classList.remove('liked');
          } else {
            e.currentTarget.classList.add('liked');
          }
          handleLike(postId, userId, () => {
            // CallBack después de un like
            const userPostsCollection = collection(firestore, 'post');
            getDocs(userPostsCollection).then((snapshot) => {
              setupPost(snapshot.docs);
            });
          });
        });
      });
      // Evento Delete
      const deleteButton = document.querySelectorAll('.deleteButton');
      deleteButton.forEach((button) => {
        button.addEventListener('click', (e) => {
          // eslint-disable-next-line
           const alertDelete = confirm('¿Está segur@ que desea eliminar este post?');
          const postId = e.currentTarget.getAttribute('data-post-id');
          if (alertDelete === true) {
            deletePost(postId);
            alert('Post eliminado con éxito');
          } else {
            alert('Operación cancelada');
          }
        });
      });
      // Evento Editar
      const editButtons = document.querySelectorAll('.editButton');
      editButtons.forEach((button) => {
        button.addEventListener('click', (e) => {
          const postId = e.currentTarget.getAttribute('data-post-id');
          const textareaTitle = document.querySelector(
            `.editTextarea[data-post-id="${postId}"]`,
          );
          const textareaDescription = document.querySelector(
            `.editContentTextarea[data-post-id="${postId}"]`,
          );
          const saveEditButton = document.querySelector(
            `.saveEditButton[data-post-id="${postId}"]`,
          );
          const likeButton = document.querySelector(
            `.containerLikes[data-post-id="${postId}"]`,
          );
          const descriptionEdit = document.querySelector(
            `.editPublic[data-post-id="${postId}"]`,
          );
          textareaTitle.style.display = 'flex';
          textareaDescription.style.display = 'flex';
          saveEditButton.style.display = 'flex';
          descriptionEdit.style.display = 'flex';
          likeButton.style.display = 'none';

          saveEditButton.addEventListener('click', () => {
            editPost(postId, textareaTitle.value, textareaDescription.value)
              .then(() => {
                alert('Publicación editada con éxito');
              })
              .catch((error) => {
                console.error('Error al editar el post:', error);
              });
          });
        });
      });
    } else {
      // console.error('Data is not an array:', data);
      viewPost.innerHTML = '<p> Aun no hay publicaciones </p>';
    }
  }

  // evento cerrar sesion
  logOutIcon.addEventListener('click', () => {
    // eslint-disable-next-line
    const alertlogOut = confirm('¿Está segur@ que desea salir de su cuenta?');
    if (alertlogOut === true) {
      // mainPage.removeAttribute('class', 'homepage')
      logOut(navigateTo);
    } else {
      alert('Operación cancelada');
    }
  });

  initializeAuth(setupPost);

  mainPage.append(backgroundLayer, headerPost, containerPubication, viewPost);
  headerPost.append(logoImage, logOutIcon);
  logOutIcon.append(iconLogOut);
  containerPubication.append(imagePublication, containerPost);
  containerPost.append(postTitle, post, imageInput, buttonSave);
  return mainPage;
}
