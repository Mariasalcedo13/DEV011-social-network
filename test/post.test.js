import { posts } from '../src/post.js';
import * as firebase from '../src/firebase.js';
import renderLogin from '../src/login.js';

describe('Usuario autenticado', () => {
  it('Inicia sesión correctamente', () => {
    // Mock de la función onAuthStateChanged de firebase
    jest.spyOn(firebase, 'initializeAuth').mockImplementation((callback) => {
      // Simula un usuario no autenticado
      callback(null);
    });
    // Mock de la función de navegación
    const navigateTo = jest.fn();

    // Mock de la función signInWithEmailAndPassword de firebase
    jest.spyOn(firebase, 'login').mockResolvedValueOnce({});

    // Renderiza el componente
    const mainPage = document.createElement('div');
    mainPage.append(renderLogin(navigateTo));
    document.body.append(mainPage);
    // Simula el inicio de sesión
    const loginEmail = mainPage.querySelector('#emailLog');
    loginEmail.value = 'holahola@holahola.com';
    const loginPassword = mainPage.querySelector('#passwordLogin');
    loginPassword.value = 'holaholahola';
    const buttonLogin = mainPage.querySelector('#sessionBtn');
    buttonLogin.click();

    // Verifica que la función de inicio de sesión se haya llamado correctamente
    expect(firebase.login).toHaveBeenCalledTimes(1);

    // Verifica que la función de navegación se haya llamado después de iniciar sesión
    // expect(navigateTo).toHaveBeenCalledWith('/posts');
  });
});

describe('posts', () => {
  it('Should be a function', () => {
    expect(typeof posts).toBe('function');
  });
  it('Have a button to log out', () => {
    const DOM = document.createElement('div');
    document.body.appendChild(DOM);
    DOM.append(posts());
    const haveAbutton = DOM.querySelector('.logOutButton');
    expect(haveAbutton).not.toBe(undefined);
  });
  it('Have a container to posts', () => {
    const DOM = document.createElement('div');
    DOM.append(posts());
    const containerPosts = DOM.querySelector('.postView');
    expect(containerPosts).not.toBe(undefined);
  });
  it('Have a form to public post', () => {
    const DOM = document.createElement('div');
    DOM.append(posts());
    const createPosts = DOM.querySelector('#task-form');
    expect(createPosts).not.toBe(undefined);
  });
  it('Post have a button to delete', () => {
    const DOM = document.createElement('div');
    DOM.append(posts());
    const buttonDelete = DOM.querySelector('.deleteButton');
    expect(buttonDelete).not.toBe(undefined);
  });
  it('Post have a button to edit', () => {
    const DOM = document.createElement('div');
    document.body.append(DOM);
    DOM.append(posts());
    const buttonEdit = DOM.querySelector('.editButton');
    expect(buttonEdit).not.toBe(undefined);
  });
  it('Snapshot of posts', () => {
    const DOM = document.createElement('div');
    DOM.append(posts());
    expect(DOM).toMatchSnapshot();
  });
});
