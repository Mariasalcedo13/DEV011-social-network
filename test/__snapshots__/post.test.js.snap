// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`posts Snapshot of posts 1`] = `
<div>
  <div
    class="homepage"
    id="homepagePostes"
  >
    <div
      class="background-layer"
    />
    <div
      class="headerPost"
    >
      Mi Plantapp
      <img
        class="logoImage"
        src="/img/planta-arana.png"
      />
      <button
        class="logOutButton"
      >
        <img
          src="/img/salir.png"
        />
      </button>
    </div>
    <div
      class="containerPubication"
    >
      <img
        class="imagePublication"
        src="/img/mujer.png"
      />
      <form
        id="task-form"
      >
        <input
          class="postTitle"
          placeholder="Título de la publicación."
          type="text"
        />
        <textarea
          id="postText"
          placeholder="Ingresa el contenido de la publicación."
        />
        <input
          accept="image/*"
          id="post-image"
          type="file"
        />
        <button
          class="buttonSave"
        >
          Publicar
        </button>
      </form>
    </div>
    <div
      class="postView"
    >
      <p>
         Aun no hay publicaciones 
      </p>
    </div>
  </div>
</div>
`;
